include/seastar/util/log.hh

```diff
@@ -120,6 +123,53 @@ public:
 
     /// \cond internal
     /// \brief used to hold the log format string and the caller's source_location.
+#ifdef SEASTAR_LOGGER_COMPILE_TIME_FMT
+    template<typename... Args>
+    struct format_info {
+        /// implicitly construct format_info from a constant format string
+        /// \param fmt - {fmt} style format string
+        template<
+            typename S,
+            std::enable_if_t<std::is_convertible_v<const S&, std::string_view>, int> = 0>
+        FMT_CONSTEVAL inline format_info(const S& format,
+                           compat::source_location loc = compat::source_location::current()) noexcept
+            : format(format)
+            , loc(loc)
+        {}
+        inline format_info(fmt::format_string<Args...> s,
+                           compat::source_location loc = compat::source_location::current()) noexcept
+            : format(s)
+            , loc(loc)
+        {}
+#if FMT_VERSION >= 10'00'00
+        using runtime_format_string_t = fmt::runtime_format_string<char>;
+#else
+        using runtime_format_string_t = fmt::basic_runtime<char>;
+#endif
+        inline format_info(runtime_format_string_t s,
+                           compat::source_location loc = compat::source_location::current()) noexcept
+            : format(s)
```

Comment:
I'm surprised it's legal to copy a runtime_format_string into a format_string<Args...>. What happens if the types don't match? Does it throw during the copy, or during formatting? If the former, will our try/catch capture the exception? (I think not).

Reply:
tl;dr: it throws in `fmt::format_to()`.

long story: `format_string<Args...>` is an alias of `basic_format_string<char, type_identity_t<Args>...>`, see https://github.com/fmtlib/fmt/blob/50565f9853926501bd1c9bda07c6a98f4d940691/include/fmt/core.h#L2794-L2795 , and `basic_format_string<...>` can be constructed from a `runtime_format_string<char>`, see https://github.com/fmtlib/fmt/blob/50565f9853926501bd1c9bda07c6a98f4d940691/include/fmt/core.h#L2783 .

because the compile-time format check / parse only applies to the `format_string` 's another constructor, see https://github.com/fmtlib/fmt/blob/50565f9853926501bd1c9bda07c6a98f4d940691/include/fmt/core.h#L2772-L2781. if the `Args` fail to parse, our  try/catch would indeed capture the exception.

actually, we do have a test for this case, see https://github.com/scylladb/seastar/blob/550da0e6750a625a171934240c9f7842ab05bc58/tests/unit/exception_logging_test.cc#L186. and in this change, it is updated to pass a `runtime_format_string<>` to it, see https://github.com/scylladb/seastar/pull/1196/files/eca4fb1f2264057d7fd3052942746b44a7064f20..4e2185f6af34245ddea89afe08114bd58525cf1d#diff-88cea8dac82294af696cf8e977984ec364ac29944c96cb18e8339e15b10498b5R194 . where the format string includes a placeholder, but we intentionally pass no args to be formatted. so a `fmt::format_error` is throw by `fmt::format_to()`, and we catch `...` in `logger::log()`, so this is captured and logged. 

FWIW, the logging message reads:
```
/home/kefu/dev/seastar/tests/unit/exception_logging_test.cc:194 @void format_error_test::test_method(): failed to log message: fmt='bad format string: {}': fmt::v10::format_error (argument not found)
```
in my test env.

Reply:
ok. It's surprising, because it means that format checking must be done twice: Once during compile time, then again during run-time, since the formatter has no way of knowing where the string came from. It also cannot pre-parse the string by calling formatter::parse during compile time (if they're constexpr).

Reply:
yeah, kind of. so two passes

1. in the ctor: it's like a dry-run -- `format_string_checker` is passed to `parse_format_string`
2. in `vformat_to()`: it actually performs the format -- `format_handler` is passed to `parse_format_string`